### DISCUSSION TITLE
Suggestions for Most Accurate State Data

### DISCUSSION THREAD

**Discussion by eli-pratt** [bitcraze/discussions#1502] [Q&A]
Hello!

I am trying to get precise position, orientation, linear velocity, and angular velocity data for autonomous flight. I have access to a Vicon motion capture system and have been able to get good data from it. I also tried out the drone's IMU sensors by logging orientation data, and I noticed that I would get some yaw drift over time. I saw that another person had this issue in #647 that explained that yaw was more prone to drift when only using IMU. 

Here is orientation data from the Crazyflie when it was sitting still on a desk only using IMU. The yaw continued to increase over time and never went back down.

<img width="935" alt="Screen Shot 2024-08-02 at 12 22 20 PM" src="https://github.com/user-attachments/assets/0dfd7f3d-0ee9-4891-a088-d9e56b414ecf">

So, to get the data I need, would you suggest that I fuse the Vicon data and IMU data or just use the Vicon data and ignore the IMU readings?

**Comment by knmcguire** [comment]
Hi!

We have seen some reports of this but unfortunately it is difficult to replicate. Could you answer a couple of questions on this?
* Have you started up the crazyflie while it was flat on a table, or at an angle.
* At this moment that the yaw starts to diverge, was the crazyflie receiving any vicon data already?
* Which estimator was selected? The complimentary filter or the EKF

**Reply by eli-pratt** [reply]
Hello!

Thanks for the reply!

- I started up the crazyflie when it was flat on a table.
- The crazyflie was not receiving any Vicon data here; this was only using IMU.
- The complementary filter was selected.

**Reply by knmcguire** [reply]
Do you see the same thing happening when the EKF was selected? 

Just a headsup that the complementary filter does not do any thing with position estimates (only height I think), so if you do feed it vicon information, make sure to select the EKF.

**Reply by eli-pratt** [reply]
<img width="909" alt="Screen Shot 2024-08-07 at 12 12 12 PM" src="https://github.com/user-attachments/assets/3eb61d9a-1d80-4b46-96d9-99fd647d62c0">

This is what the data looks like when using the EKF. The yaw value is now decreasing as time goes on. 

<img width="993" alt="Screen Shot 2024-08-07 at 12 10 39 PM" src="https://github.com/user-attachments/assets/357f4686-0b78-4f98-a311-d83c7aedf717">

I also tried logging the state estimate roll, pitch, and yaw variables and it also gave me weird data where it jumped from 0.0006 to 6 in 200 ms.

**Reply by knmcguire** [reply]
Hmmm I don't see the -6 myself, perhaps you are referring to the -6e-5 value in yaw? that is just a formatting feature. 

You should perhaps try to look at plots instead of readouts to have a better view than this. Either you can use the rqt plotter if you want to view it in ROS or use the cfclient's plotter functionality: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/plotter_tab/

And about the state-estimate yaw diverging, state estimation based on IMU alone is tricky business! The output from the gyroscope is angle rate and not absolute, so technically it's doing dead reckoning. At least the roll and pitch estimate can still relate to the acceleratometer (to know where the gravity factor is pointing too) but the yaw estimate can't rely on that and only has gyro. So sometimes when the situation is ideal, it works well enough to not drift too much for the duration of the crazyflie's battery but a lot of times it doesn't. At least with the EKF, it also has help of the dynamics model of the crazyflie itself so usually the yaw estimate improves a lot in flight. But sometimes it's just not enough for it to converge if there is a lot of noise, or if the crazyflie is initialized at a small angle (lot's of factors can tip of the delicate balance). If you add position to it from an external positioning system then it has more information to go with and usually that is enough for the EKF to converge.

Hope that that explains it?
